ReactiveObjC/RACSignal+Operations.m

```diff
@@ -846,22 +846,57 @@ - (RACSignal *)takeUntilReplacement:(RACSignal *)replacement {
 
 - (RACSignal *)switchToLatest {
   return [[RACSignal createSignal:^(id<RACSubscriber> subscriber) {
-    RACMulticastConnection *connection = [self publish];
+    __block BOOL outerCompleted = NO;
+    __block BOOL innerCompleted = NO;
+    __block NSUInteger innerSignalIndex = 0;
 
-    RACDisposable *subscriptionDisposable = [[connection.signal
-      flattenMap:^(RACSignal *x) {
-        NSCAssert(x == nil || [x isKindOfClass:RACSignal.class], @"-switchToLatest requires that the source signal (%@) send signals. Instead we got: %@", self, x);
+    RACSerialDisposable *innerDisposable = [[RACSerialDisposable alloc] init];
+    NSLock *lock = [[NSLock alloc] init];
+
+    RACDisposable *outerDisposable = [self subscribeNext:^(RACSignal *signal) {
+      NSCAssert(signal == nil || [signal isKindOfClass:RACSignal.class], @"-switchToLatest expects signals as values, instead we got: %@", signal);
+      signal = signal ?: [RACSignal empty];
 
-        // -concat:[RACSignal never] prevents completion of the receiver from
-        // prematurely terminating the inner signal.
-        return [x takeUntil:[connection.signal concat:[RACSignal never]]];
-      }]
-      subscribe:subscriber];
+      [innerDisposable.disposable dispose];
+
+      [lock lock];
+      innerCompleted = NO;
+      ++innerSignalIndex;
+      NSUInteger signalIndex = innerSignalIndex;
+      [lock unlock];
+
+      RACDisposable *disposable = [signal subscribeNext:^(id x) {
+        [subscriber sendNext:x];
+      } error:^(NSError *error) {
+        [subscriber sendError:error];
+      } completed:^{
+        [lock lock];
+        if (innerSignalIndex == signalIndex) {
```

Comment:
All of this index shenanigans is only to overcome the potential sending of values after disposal?

Reply:
Without this, it's possible for the previous inner signal to complete and then if the outer signal completes, the signal will complete. This is although there's a new inner signal which is still active.

I know it's a bit clunky. Let me know if you have a more elegant solution here.

Reply:
Yes but you update the index after you dispose the previous signal. So my question is: can it still potentially send values after disposal?

Reply:
I see that's what I figured.

Wouldn't it be more straightforward to store the pointer of the current active inner signal?

Reply:
Glad you asked! This may not work since no one guarantees that pointers will not be reused in another allocation.

We may be able to hold the signal weakly and then compare, though. Let me know if you find any loopholes here.

Reply:
Can you point me in the direction of a trustworthy resource regarding pointer reusability? Couldn't find anything. 

If @Shachlan is stressed about this thing I can let this counter blasphemy slide.

Reply:
Yes, this delays a version. I'd like it to be merged quickly.

Reply:
I don't know any source that explains it, but it's very simple - the signal you hold the pointer of can get deallocated (since it's not being held by ARC) and then the pointer can get reassigned to a different object. By comparing pointers you don't really know that one of the pointers got replaced with a completely different object.

Using `__weak` should solve this as the original pointer will get set to `nil`.

Reply:
So whichever you prefer. If Schahar wants it asap the counters will also suffice.

Reply:
The time pressure is off - we've circumvented the issue,  so you can finish the PR properly.

Reply:
Also, kudos on the proper Deutsche spelling of my name.